lightning/src/ln/channelmanager.rs

```diff
@@ -2289,10 +2346,38 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 								let mut payment_secrets = self.pending_inbound_payments.lock().unwrap();
 								match payment_secrets.entry(payment_hash) {
 									hash_map::Entry::Vacant(_) => {
-										log_trace!(self.logger, "Failing new HTLC with payment_hash {} as we didn't have a corresponding inbound payment.", log_bytes!(payment_hash.0));
-										fail_htlc!(claimable_htlc);
+										match claimable_htlc.onion_payload {
+											OnionPayload::Invoice(_) => {
+												log_trace!(self.logger, "Failing new HTLC with payment_hash {} as we didn't have a corresponding inbound payment.", log_bytes!(payment_hash.0));
+												fail_htlc!(claimable_htlc);
+											},
+											OnionPayload::Spontaneous(preimage) => {
```

Comment:
Just to be clear, this doesn't prevent honest keysend replay, where a buggy/hazardous payer sends twice a keysend with the same hash, even with different payment paths, and as such opens the payment to be claimed in-flight by an intermediary node ? Note, I think this is already a risk for any hashlock-based LN payment (variant of wormhole attacks...)

I would say we even care less that w.r.t to regular payments as we keysend the preimage doesn't have the proof-of-payment property plead most of the time ?

Reply:
Hmm, yes it does seem like keysend requires more thought than normal payments for replay situations. I don't think it would open up buggy clients to in-flight claims by an intermediate node though, because wouldn't the buggy client run into issues once they started "replaying" the whole `update_add` dance, preventing the HTLC from being successfully added to the commitment tx to begin with?

Reply:
I mean the HTLC would appear under a new `htlc_id`. I'm thinking a situation similar to `duplicate_htlc_test`/`test_duplicate_htlc_different_direction_onchain`. Though nevermind, we don't prevent double-spend already for normal payments, see comment around `send_payment`.